source/extensions/omni.isaac.lab/test/controllers/test_operational_space.py

```diff
@@ -0,0 +1,864 @@
+        ee_lin_vel_b = quat_rotate_inverse(robot.data.root_quat_w, relative_vel_w[:, 0:3])  # From world to root frame
+        ee_ang_vel_b = quat_rotate_inverse(robot.data.root_quat_w, relative_vel_w[:, 3:6])
+        ee_vel_b = torch.cat([ee_lin_vel_b, ee_ang_vel_b], dim=-1)
+
+        # Calculate the contact force
+        ee_force_w = torch.zeros(self.num_envs, 3, device=self.sim.device)
+        if self.contact_forces is not None:  # Only modify if it exist
+            sim_dt = self.sim.get_physics_dt()
+            self.contact_forces.update(sim_dt)  # update contact sensor
+            # Calculate the contact force by averaging over last four time steps (i.e., to smoothen) and
+            # taking the max of three surfaces as only one should be the contact of interest
+            ee_force_w, _ = torch.max(torch.mean(self.contact_forces.data.net_forces_w_history, dim=1), dim=1)
+
+        # This is a simplification, only for the sake of testing.
+        ee_force_b = ee_force_w
+
+        return jacobian_b, mass_matrix, gravity, ee_pose_b, ee_vel_b, root_pose_w, ee_pose_w, ee_force_b
+
+    def _update_target(
+        self,
+        osc: OperationalSpaceController,
+        root_pose_w: torch.tensor,
+        ee_pose_b: torch.tensor,
+        target_set: torch.tensor,
+        current_goal_idx: int,
+    ):
+        # update the ee desired command
+        command = torch.zeros(self.num_envs, osc.action_dim, device=self.sim.device)
+        command[:] = target_set[current_goal_idx]
+
+        # update the ee desired pose
+        ee_target_pose_b = torch.zeros(self.num_envs, 7, device=self.sim.device)
+        for target_type in osc.cfg.target_types:
+            if target_type == "pose_abs":
+                ee_target_pose_b[:] = command[:, :7]
+            elif target_type == "pose_rel":
+                ee_target_pose_b[:, 0:3], ee_target_pose_b[:, 3:7] = apply_delta_pose(
+                    ee_pose_b[:, :3], ee_pose_b[:, 3:], command[:, :7]
+                )
+            elif target_type == "wrench_abs":
+                pass  # ee_target_pose_b could stay at the root frame for force control, what matters is ee_target_b
+            else:
+                raise ValueError("Undefined target_type within _update_target().")
+
+        # update the target desired pose in world frame (for marker)
+        ee_target_pos_w, ee_target_quat_w = combine_frame_transforms(
+            root_pose_w[:, 0:3], root_pose_w[:, 3:7], ee_target_pose_b[:, 0:3], ee_target_pose_b[:, 3:7]
+        )
+        ee_target_pose_w = torch.cat([ee_target_pos_w, ee_target_quat_w], dim=-1)
+
+        next_goal_idx = (current_goal_idx + 1) % len(target_set)
+
+        return command, ee_target_pose_b, ee_target_pose_w, next_goal_idx
+
+    def _convert_to_task_frame(
+        self, osc: OperationalSpaceController, command: torch.tensor, ee_target_pose_b: torch.tensor
+    ):
+        command = command.clone()
```

Comment:
please add docstrings for this function

Reply:
"""Convert the target command to the task frame if required.

        Args:
            osc (OperationalSpaceController): The operational space controller.
            command (torch.tensor): The target command to convert.
            ee_target_pose_b (torch.tensor): The end-effector target pose in the body frame.

        Returns:
            command (torch.tensor): The converted target command.
            task_frame_pose_b (torch.tensor): The task frame pose in the body frame.

        Raises:
            ValueError: If the frame is invalid.
        """